daemon-tests/src/flow.rs

```diff
@@ -21,15 +21,25 @@ pub async fn next_maker_offers(
     let mut rx_b = rx_b.clone();
 
     let non_empty_offer = |offer: MakerOffers| {
-        if offer.long.is_some()
-            && offer.short.is_some()
-            && contract_symbol == &offer.long.clone().unwrap().contract_symbol
-            && contract_symbol == &offer.short.clone().unwrap().contract_symbol
-        {
-            Some(offer)
-        } else {
-            None
+        match &offer.long {
+            Some(long) => {
+                if contract_symbol != &long.contract_symbol {
+                    return None;
+                }
+            }
+            None => return None,
         }
+
+        match &offer.short {
+            Some(short) => {
+                if contract_symbol != &short.contract_symbol {
+                    return None;
+                }
+            }
+            None => return None,
+        }
+
+        Some(offer)
```

Comment:
🙃 It's a nice idea to use pattern matching here, but you can even go a bit further. Let me show you something that I consider more Rust-idiomatic:

```rust
let non_empty_offers = |offers: MakerOffers| match &offers {
    MakerOffers {
        long: Some(long),
        short: Some(short),
    } if &long.contract_symbol == contract_symbol
        && &short.contract_symbol == contract_symbol =>
    {
        Some(offers)
    }

    _ => None,
};
```

There are a couple things going on here that might be interesting:

- Pattern matching against a custom struct, which can be really powerful.
- Match [guards](https://doc.rust-lang.org/rust-by-example/flow_control/match/guard.html).
- Avoiding explicit returns. I think idiomatic Rust uses returns sparingly. I'd say this team strives to use them mostly in early returns or not at all. The reason being that it's easier to read a function when you can expect it to return from only one or two predictable places.

Reply:
thank you for that rust lesson 😃